### PR TITLE
[SPARK-33836][SS][PYTHON] Expose DataStreamReader.table and DataStreamWriter.toTable

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to expose `DataStreamReader.table` (SPARK-32885) and `DataStreamWriter.toTable` (SPARK-32896) to PySpark, which are the only way to read and write with table in Structured Streaming.

### Why are the changes needed?

Please refer SPARK-32885 and SPARK-32896 for rationalizations of these public APIs. This PR only exposes them to PySpark.

### Does this PR introduce _any_ user-facing change?

Yes, PySpark users will be able to read and write with table in Structured Streaming query.

### How was this patch tested?

Manually tested.

> v1 table

>> create table A and ingest to the table A

```
spark.sql("""
create table table_pyspark_parquet (
    value long,
    `timestamp` timestamp
) USING parquet
""")
df = spark.readStream.format('rate').option('rowsPerSecond', 100).load()
query = df.writeStream.toTable('table_pyspark_parquet', checkpointLocation='/tmp/checkpoint5')
query.lastProgress
query.stop()
```

>> read table A and ingest to the table B which doesn't exist

```
df2 = spark.readStream.table('table_pyspark_parquet')
query2 = df2.writeStream.toTable('table_pyspark_parquet_nonexist', format='parquet', checkpointLocation='/tmp/checkpoint2')
query2.lastProgress
query2.stop()
```

>> select tables

```
spark.sql("DESCRIBE TABLE table_pyspark_parquet").show()
spark.sql("SELECT * FROM table_pyspark_parquet").show()

spark.sql("DESCRIBE TABLE table_pyspark_parquet_nonexist").show()
spark.sql("SELECT * FROM table_pyspark_parquet_nonexist").show()
```

> v2 table (leveraging Apache Iceberg as it provides V2 table and custom catalog as well)

>> create table A and ingest to the table A

```
spark.sql("""
create table iceberg_catalog.default.table_pyspark_v2table (
    value long,
    `timestamp` timestamp
) USING iceberg
""")
df = spark.readStream.format('rate').option('rowsPerSecond', 100).load()
query = df.select('value', 'timestamp').writeStream.toTable('iceberg_catalog.default.table_pyspark_v2table', checkpointLocation='/tmp/checkpoint_v2table_1')
query.lastProgress
query.stop()
```

>> ingest to the non-exist table B

```
df2 = spark.readStream.format('rate').option('rowsPerSecond', 100).load()
query2 = df2.select('value', 'timestamp').writeStream.toTable('iceberg_catalog.default.table_pyspark_v2table_nonexist', checkpointLocation='/tmp/checkpoint_v2table_2')
query2.lastProgress
query2.stop()
```

>> ingest to the non-exist table C partitioned by `value % 10`

```
df3 = spark.readStream.format('rate').option('rowsPerSecond', 100).load()
df3a = df3.selectExpr('value', 'timestamp', 'value % 10 AS partition').repartition('partition')
query3 = df3a.writeStream.partitionBy('partition').toTable('iceberg_catalog.default.table_pyspark_v2table_nonexist_partitioned', checkpointLocation='/tmp/checkpoint_v2table_3')
query3.lastProgress
query3.stop()
```

>> select tables

```
spark.sql("DESCRIBE TABLE iceberg_catalog.default.table_pyspark_v2table").show()
spark.sql("SELECT * FROM iceberg_catalog.default.table_pyspark_v2table").show()

spark.sql("DESCRIBE TABLE iceberg_catalog.default.table_pyspark_v2table_nonexist").show()
spark.sql("SELECT * FROM iceberg_catalog.default.table_pyspark_v2table_nonexist").show()

spark.sql("DESCRIBE TABLE iceberg_catalog.default.table_pyspark_v2table_nonexist_partitioned").show()
spark.sql("SELECT * FROM iceberg_catalog.default.table_pyspark_v2table_nonexist_partitioned").show()
```

### CODE REVIEW

File: python/pyspark/sql/streaming.py
Comment (HeartSaVioR) [inline python/pyspark/sql/streaming.py:1502]: Ideally it'd be nice to ship the change in 3.1.0 so that the API is available for both at the same time. I just safely set this to 3.2.0 to see the voices on when to add.

File: python/pyspark/sql/streaming.py
Comment (HeartSaVioR) [inline python/pyspark/sql/streaming.py:966]: Same here

File: python/pyspark/sql/streaming.pyi
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.pyi:197]: cc @zero323 FYI

**Review by dongjoon-hyun — COMMENTED** [review]
I'm also +1 for backporting this to `branch-3.1`.

cc @cloud-fan and @gatorsmile

File: python/pyspark/sql/streaming.pyi
Comment (zero323) [reply python/pyspark/sql/streaming.pyi:197]: Thanks for the ping and working on this @HyukjinKwon!

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres @viirya @gaborgsomogyi @xuanyuanking

File: python/pyspark/sql/streaming.py
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.py:1535]: @HeartSaVioR I think we should avoid using `/tmp` directory directly, see also SPARK-24126. Probably we should use `tempfile.mkdtemp()` instead.

File: python/pyspark/sql/streaming.py
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.py:1593]: Can we remove these stuff, make the docstring example pretty, and add a proper unittest into `sql/tests/test_streaming.py`?

File: python/pyspark/sql/streaming.py
Comment (HyukjinKwon) [reply python/pyspark/sql/streaming.py:1502]: I am okay if other people are fine.

File: python/pyspark/sql/streaming.py
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.py:957]: I think it's fine to just copy and paste Scaladoc, and document the return type as below according to numpydoc style

```
Define a Streaming DataFrame on a Table. The DataSource corresponding to the table should
support streaming mode.

Parameters
...

Returns
--------
:class:`DataFrame`

```

File: python/pyspark/sql/streaming.py
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.py:1500]: This one as well. I think it's fine to match the docs with Scaladoc. 

btw nit `[[StreamingQuery]]` -> `` :class:`StreamingQuery` ``

**Comment by HyukjinKwon** [comment]
Looks pretty good

File: python/pyspark/sql/streaming.py
Comment (HeartSaVioR) [reply python/pyspark/sql/streaming.py:957]: I saw functions here have their own description so I thought we prefer re-describing it. OK I'll copy the description. Thanks.

Comment (HeartSaVioR) [reply python/pyspark/sql/streaming.py:957]: Just to make sure, could you please guide about the content of doc in this method? I'm not sure whether you're suggesting to delete something or just rearrange and make a sync with Scala doc.

I'll follow the same with other method once you provide the example. Thanks!

Comment (HyukjinKwon) [reply python/pyspark/sql/streaming.py:957]: Ohh, I was clear. I meant, for example,

```python
"""Define a Streaming DataFrame on a Table. The DataSource corresponding to the table should
support streaming mode.

.. versionadded:: 3.2.0

Parameters
----------
tableName : str
    string, for the name of the table.

Returns
--------
:class:`DataFrame`

Notes
-----
This API is evolving.

Examples
--------
>>> csv_sdf = spark.readStream.table(input_table_name)
>>> csv_sdf.isStreaming
True
"""
```

File: python/pyspark/sql/streaming.py
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.py:1551]:
```python
        """
        Starts the execution of the streaming query, which will continually output results to the given
        table as new data arrives. A new table will be created if the table not exists. The returned
        :class:`StreamingQuery` object can be used to interact with the stream.
   
        .. versionadded:: 3.2.0

        Parameters
        ----------
        tableName : str
            string, for the name of the table.
        format : str, optional
            the format used to save.
        outputMode : str, optional
            specifies how data of a streaming DataFrame/Dataset is written to a
            streaming sink.

            * `append`: Only the new rows in the streaming DataFrame/Dataset will be written to the
              sink
            * `complete`: All the rows in the streaming DataFrame/Dataset will be written to the
              sink every time these is some updates
            * `update`: only the rows that were updated in the streaming DataFrame/Dataset will be
              written to the sink every time there are some updates. If the query doesn't contain
              aggregations, it will be equivalent to `append` mode.
        partitionBy : str or list, optional
            names of partitioning columns
        queryName : str, optional
            unique name for the query
        **options : dict
            All other string options. You may want to provide a `checkpointLocation`.

        Notes
        -----
        This API is evolving.

        Examples
        --------
        >>> sq = sdf.writeStream.format('parquet').queryName('this_query').option(
        ...      'checkpointLocation', '/tmp/checkpoint').toTable(output_table_name)
        >>> sq.isActive
        True
        >>> sq.name
        'this_query'
        >>> sq.stop()
        >>> sq.isActive
        False
        >>> sq = sdf.writeStream.trigger(processingTime='5 seconds').toTable(
        ...     output_table_name, queryName='that_query', outputMode="append", format='parquet',
        ...     checkpointLocation='/tmp/checkpoint')
        >>> sq.name
        'that_query'
        >>> sq.isActive
        True
        >>> sq.stop()
        """
```

or .. 

```python
        """
        Starts the execution of the streaming query, which will continually output results to the given
        table as new data arrives.
   
        A new table will be created if the table not exists. The returned
        :class:`StreamingQuery` object can be used to interact with the stream.
   
        .. versionadded:: 3.2.0

        Parameters
        ----------
        tableName : str
            string, for the name of the table.
        format : str, optional
            the format used to save.
        outputMode : str, optional
            specifies how data of a streaming DataFrame/Dataset is written to a
            streaming sink.

            * `append`: Only the new rows in the streaming DataFrame/Dataset will be written to the
              sink
            * `complete`: All the rows in the streaming DataFrame/Dataset will be written to the
              sink every time these is some updates
            * `update`: only the rows that were updated in the streaming DataFrame/Dataset will be
              written to the sink every time there are some updates. If the query doesn't contain
              aggregations, it will be equivalent to `append` mode.
        partitionBy : str or list, optional
            names of partitioning columns
        queryName : str, optional
            unique name for the query
        **options : dict
            All other string options. You may want to provide a `checkpointLocation`.

        Notes
        -----
        This API is evolving.

        Examples
        --------
        >>> sq = sdf.writeStream.format('parquet').queryName('this_query').option(
        ...      'checkpointLocation', '/tmp/checkpoint').toTable(output_table_name)
        >>> sq.isActive
        True
        >>> sq.name
        'this_query'
        >>> sq.stop()
        >>> sq.isActive
        False
        >>> sq = sdf.writeStream.trigger(processingTime='5 seconds').toTable(
        ...     output_table_name, queryName='that_query', outputMode="append", format='parquet',
        ...     checkpointLocation='/tmp/checkpoint')
        >>> sq.name
        'that_query'
        >>> sq.isActive
        True
        >>> sq.stop()
        """
```

Either way should be fine. There's no strict rule on the docstrings currently (except https://numpydoc.readthedocs.io/en/latest/format.html). It's just my suggestion :-).

File: python/pyspark/sql/streaming.py
Comment (viirya) [reply python/pyspark/sql/streaming.py:1502]: I think it is okay as this only expose the APIs to PySpark.

**Review by viirya — COMMENTED** [review]
Looks okay to me.

File: python/pyspark/sql/streaming.py
Comment (viirya) [inline python/pyspark/sql/streaming.py:1517]: there is -> there are?

File: python/pyspark/sql/streaming.py
Comment (viirya) [reply python/pyspark/sql/streaming.py:1593]: +1

**Comment by HeartSaVioR** [comment]
I see two explicit supports on landing this to 3.1.0 and no negative input. Let me set the version to 3.1.0 as of now. I'll revert it once we decide to not include this in 3.1.0 in later review phase.

**Review by HyukjinKwon — APPROVED** [review]
Lgtm. I have some minor nits and suggestions on my mind but that shouldn't block this PR. I'll follow up by myself if you don't mind @HeartSaVioR.

**Comment by HyukjinKwon** [comment]
I'll merge this on Monday in KST if there are no outstanding comments.

**Comment by HeartSaVioR** [comment]
Yeah sure, thanks for the detailed review, @HyukjinKwon !

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Review by xuanyuanking — COMMENTED** [review]
Late LGTM.
I'll submit PR for SPARK-33659 today. Sorry for the late. (Some other thing blocked me in the last 2 weeks...)

**Pull request closed without merging**
